src/events/eventActions.js

```diff
@@ -320,8 +319,4 @@ export const deadQueue = (): ThunkAction<Promise<void>> => async (dispatch, getS
 
   // TODO(#3881): Lots of issues with outbox sending
   dispatch(sendOutbox());
-
-  // TODO: Remove this. No need to do this on DEAD_QUEUE or ACCOUNT_SWITCH:
-  //   https://chat.zulip.org/#narrow/stream/48-mobile/topic/.60remoteNotificationsRegistered.60.20listener.20too.20late/near/1148915
-  dispatch(initNotifications());
```

Comment:
Remind me: was there something that actually goes wrong as a result of this, that removing it fixes? Or was it just a matter of feeling that this wasn't the right place for it?

As I think more about this action, I think it's actually in a situation very much parallel to `sendOutbox`: there's some data we have (be it our notifications token, or the user's messages they've tried to send); we want to get it to the server; when we eventually succeed at that, we'll update our data to reflect that, so we always have a to-do list of what we still want to send; and we want to keep trying in order to get it through to the server, until we succeed (or in the case of a message, perhaps time out.)

Those, and: we don't currently have a real principled way of driving those retries. Instead, we take the occasion of the initial fetch as a hook to get us to retry. It'd be good to get a principled way of doing them (for outbox, that's #3881), but until then every place we retry is potentially a helpful one, even if ad hoc.

So if there isn't a concrete problem that's solved by removing these `initNotification` calls, I think I'd prefer to actually keep them. After #3881, we can take whatever we work out for that and make something similar for driving these retries too.

Reply:
> Remind me: was there something that actually goes wrong as a result of this, that removing it fixes? Or was it just a matter of feeling that this wasn't the right place for it?

Yeah, as far as I remember it's the latter. I'll adjust.

Reply:
Yeah, we'll have to think carefully about retries.

Even setting that aside, I've often found our code for registering push tokens pretty hard to follow, and wondered if we could make it better.

We quick-fixed a P0 bug #4573 with 6ffc7e18a, and I don't think we've settled on a plan for something more permanent.

I think this is the description of what the relevant subsystem is aimed at:

For all servers we want notifications from, we should keep the server up-to-date with our current device token.
- If we're logged into a server, we want notifications from it.
  - This suggests having some per-account code that runs with the account's LOGIN_SUCCESS.
- Our device token could change between sessions of running the app, or during a session.
  - This suggests having some global code that runs on startup, and on learning (e.g., from an event) that the token has changed during this session. Global because the token is for the device, and also because we may need to tell multiple servers about the token.

I think there's a main reason why our code has developed such that it's not super clear how we answer these requirements: we're not supposed to ask iOS for the token until we're ready to pop up a permissions request, and we're supposed to do that only in a context where the permission would be useful to have. We've (reasonably) decided that the app enters that context when a `/register` completes, so that's when we request the token on iOS.

Here's an example of how the code is hard to follow. It occurred to me after your comment prompted me to keep thinking about this:

> - Our device token could change between sessions of running the app, or during a session.
>   - This suggests having some global code that runs on startup […]. Global because the token is for the device, and also because we may need to tell multiple servers about the token.

Our "on-startup" code for this is `initNotifications`. That sure doesn't look like "some global code", does it? It's a `ThunkAction`, not a `GlobalThunkAction`. But then, if you sit and think about it, and read some distant code, it's possible to convince yourself that all interested servers will indeed be notified if the device token changed between sessions of running the app. Better if the "on-startup" code were instead a `GlobalThunkAction`, with a nice jsdoc that mentioned that device tokens can change between sessions.

[I have a draft](https://chat.zulip.org/#narrow/stream/48-mobile/topic/.60remoteNotificationsRegistered.60.20listener.20too.20late/near/1147424) from a year ago for a unified `getAndReturnNotificationToken` function that grabs the token in the platform-appropriate way. That specific draft may or may not end up being helpful, but I think if we chose one of the following, we could unify our separate iOS/Android code and thus remove lots of complexity:

1. iOS and Android both get the token immediately on startup (sad for iOS because the permission pop-up isn't deferred till after `/register`), or
2. iOS and Android both get the token after `/register`, where iOS gets it currently (sad for Android because this isn't necessary, but also it isn't harmful, is it?)

Reply:
> I think this is the description of what the relevant subsystem is aimed at:

Yeah, that all seems right.



> Here's an example of how the code is hard to follow. It occurred to me after your comment prompted me to keep thinking about this: […]
> 
> Our "on-startup" code for this is `initNotifications`. That sure doesn't look like "some global code", does it? It's a `ThunkAction`, not a `GlobalThunkAction`. But then, if you sit and think about it and read some distant code, it's possible to convince yourself that all interested servers will indeed be notified if the device token changed between sessions of running the app, on iOS and Android.

Hmm, indeed. I agree, that seems like a key point for why this logic is hard to follow.

Want to make an issue for this? That'll probably provide a better way to track it than this review subthread.

Reply:
Sure, sounds good: #5329.